OOG error circuit for `SLOAD` and `SSTORE`

silathdiir opened the pull request:
Close https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1119

Spec markdown PR https://github.com/privacy-scaling-explorations/zkevm-specs/pull/381

### Summary

1. Merge OOG errors for `SLOAD` and `SSTORE` into `OogError::SloadSstore`, and error states into `ExecutionState::ErrorOutOfGasSloadSstore`.

2. Move both `SloadGasGadget` and `SstoreGasGadget` to common gadget which could be reused by `SloadGadget`, `SstoreGadget` and `ErrorOOGSloadSstoreGadget`.

3. Implement `ErrorOutOfGasSloadSstore` state in both bus-mapping and zkevm-circuit.

4. Constrain `SSTORE` OOG error for both `gas_cost` and `gas_sentry`. Gas reentrancy sentry could be referenced in [this go-ethereum code](https://github.com/ethereum/go-ethereum/blob/master/core/vm/operations_acl.go#L30).

DreamWuGit commented on bus-mapping/src/evm/opcodes/error_oog_sload_sstore.rs:
```diff
@@ -0,0 +1,96 @@
+
+        let mut exec_step = state.new_step(geth_step)?;
+        let next_step = geth_steps.get(1);
+        exec_step.error = state.get_step_err(geth_step, next_step).unwrap();
```
can assign targetting error directly `exec_step.error = ExecError::OutOfGas(OogError::SloadSstore) `

silathdiir replied:
Fixed in commit [2833dd4](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1147/commits/2833dd445a76d2c2fca19ee9f089c9601948ef46).

DreamWuGit commented on zkevm-circuits/src/evm_circuit/execution/error_oog_sload_sstore.rs:
```diff
@@ -0,0 +1,585 @@
+                SLOAD
+            };
+            let mut gas_cost =
+                OpcodeId::PUSH32.constant_gas_cost().0 + cal_sload_gas_cost_for_assignment(false);
```
this line should use param  `is_warm` rather than hard code `false`?
`OpcodeId::PUSH32.constant_gas_cost().0 + cal_sload_gas_cost_for_assignment(is_warm);`

silathdiir replied:
Seems no. The first `sload(key)` is `is_warm == false` (cold key). If the parameter `is_warm == true` (warm key), then the gas cost of second `sload(key)` is `cal_sload_gas_cost_for_assignment(true)`.

The bytecode is set for `is_warm == true` as:
```
PUSH32(key)
SLOAD  // cal_sload_gas_cost_for_assignment(false)
PUSH32(key)
SLOAD  // cal_sload_gas_cost_for_assignment(true)
```

DreamWuGit replied:
I see, with a little confusion.

silathdiir commented on zkevm-circuits/src/evm_circuit/execution/error_oog_sload_sstore.rs:
```diff
@@ -0,0 +1,585 @@
+            call,
+            step,
+            // Additional one stack pop and one account storage read for SSTORE.
+            7 + if is_sstore { 2 } else { 0 },
```
Tried to fix as `7 + is_sstore * 2` but failed to compile. It seems that I set `is_sstore` as a boolean in `assign_exec_step`:
```
let is_sstore = opcode == OpcodeId::SSTORE;
```
And I use this boolean as `if is_sstore` in multiple places (otherwise it should be replaced with `if is_sstore == 1`).

silathdiir replied:
Fixed to `7 + usize::from(is_sstore) * 2` in [9fc3451](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1147/commits/9fc345150ca59c6a8e04e3d784e506b161d479ee).

DreamWuGit commented on zkevm-circuits/src/evm_circuit/execution/error_oog_sload_sstore.rs:
```diff
@@ -0,0 +1,585 @@
+        CircuitTestBuilder::new_from_test_ctx(ctx).run();
+    }
+
+    fn test_internal(call_data_offset: usize, call_data_length: usize, testing_data: &TestingData) {
```
seems `call_data_offset` and `call_data_length` don't affect test target,  suggest to remove .

silathdiir replied:
Deleted parameter `call_data_offset` and `call_data_length`, and hardcoded as `0x20` and `0x00` in `test_internal` function in [9245a20](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1147/commits/9245a20afd85c8d8d64c50e911fc76d6eacb8278).

DreamWuGit reviewed (approved):
overall LGTM

han0110 reviewed (approved):
LGTM! Just added 2 nitpicks.

Also `ExecutionState::SSTORE` seems needs to be updated to check if the gas is indeed sufficient for the `2300` (`SstoreSentryGasEIP2200`), right?

silathdiir commented:
> LGTM! Just added 2 nitpicks.
> 
> Also `ExecutionState::SSTORE` seems needs to be updated to check if the gas is indeed sufficient for the `2300` (`SstoreSentryGasEIP2200`), right?

Yes. I added a new issue https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1214. Please help assign it to me. Will fix it in a following PR. Thanks.
